File: Solutions/Python/binary_tree_maximum_path_sum.py
"""
    [HARD]
    124. Binary Tree Maximum Path Sum

    Concepts:
    - binary tree
    - depth-first search

    Stats:
        Runtime | 14 ms     [Beats 68.66%]
        Memory  | 22.92 MB  [Beats 59.94%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_path_sum = -float("inf")

        def traverse(curr: TreeNode) -> int:
            nonlocal max_path_sum

            if not curr:
                return 0

            left = max(0, traverse(curr.left))
            right = max(0, traverse(curr.right))

            # compare with path made from subtrees
            max_path_sum = max(
                max_path_sum,
                left + right + curr.val
            )

            # return max of subtrees and current value
            # returns only current value at leaf
            return max(left, right) + curr.val

        traverse(root)

        return max_path_sum
            

File: Solutions/Cpp/binary_tree_maximum_path_sum.cpp
/*
    [MEDIUM]
    124. Binary Tree Maximum Path Sum

    Concepts:
    - binary tree
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 27.82 MB  [Beats 77.25%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max_path_sum = std::numeric_limits<int>::min();

    int dfs(TreeNode* curr) {
        if (curr == nullptr)
            return 0;

        // maxes if not splitting at curr
        int left = std::max(0, dfs(curr->left));
        int right = std::max(0, dfs(curr->right));

        // compare with path made from subtrees
        max_path_sum = std::max(max_path_sum, left + right + curr->val);

        // max if splitting at curr
        return std::max(left, right) + curr->val;
    }

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return max_path_sum;
    }
};